src/main/flight/autopilot.c

```diff
@@ -91,14 +190,162 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+                acceleration = pt1FilterApply(&latLong->accelerationLpf, acceleration);
+                float pidA = acceleration * positionPidCoeffs.Kd;
+
+                // limit sum of D and A because otherwise can be too aggressive when starting at speed
+                const float maxDAAngle = 35.0f; // limit in degrees; arbitrary.  20 is a bit too low, allows a lot of overshoot
+                const float pidDA = constrainf(pidD + pidA, -maxDAAngle, maxDAAngle);
+                // note: an angle of more than 35 degrees can still be achieved as P and I grow
+
+                // ** PID Sum **
+                float pidSum = pidP + pidI + pidDA;
+
+                // reset the position target when pidSum crosses zero, typically when velocity is very close to zero, ie craft has stopped
+                // this enhances the smoothness of the transition from stick input back to position hold because there is no sharp change in pidSum
+                if (latLong->isStarting && latLong->pidSum * pidSum < 0.0f) { // pidsum ns has reversed sign
+                    resetPositionControlParams(latLong);
+                    if (i == 0) {
+                        currentTargetLocation.lat = gpsSol.llh.lat;
+                    } else {
+                        currentTargetLocation.lon = gpsSol.llh.lon;
+                    }
+                    latLong->distance = 0.0f;
+                    latLong->isStarting = false;
+                }
+                latLong->pidSum = pidSum;
+
+                // Debugs... distances in cm, angles in degrees * 10, velocities cm/2
+                if (gyroConfig()->gyro_filter_debug_axis == i) {
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 0, lrintf(distanceCm));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 1, lrintf(latLong->distance));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 2, lrintf(latLong->pidSum * 10));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 4, lrintf(pidP * 10));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 5, lrintf(pidI * 10));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 6, lrintf(pidDA * 10));
+                }
+            }
+        }
+        // ** Rotate pid Sum to quad frame of reference, into pitch and roll **
+        float headingRads = DECIDEGREES_TO_RADIANS(attitude.values.yaw);
+        const float sinHeading = sin_approx(headingRads);
+        const float cosHeading = cos_approx(headingRads);
+
+        posHold.pidSum[AI_ROLL] = -sinHeading * posHold.direction[NORTH_SOUTH].pidSum + cosHeading * posHold.direction[EAST_WEST].pidSum;
+        posHold.pidSum[AI_PITCH] = cosHeading * posHold.direction[NORTH_SOUTH].pidSum + sinHeading * posHold.direction[EAST_WEST].pidSum;
+    }
+
+    // ** Final output to pid.c Angle Mode at 100Hz with primitive upsampling**
+    autopilotAngle[AI_ROLL] = pt3FilterApply(&posHold.upsample[AI_ROLL], posHold.pidSum[AI_ROLL]);
```

Comment:
The filter will introduce delay and angular lag if quad is yawing. Filtering in earth coordinates and applying rotation  last will remove angular lag (but I'm not sure if it will improve anything).

Reply:
Its at least the correct way to do it.

Reply:
Hmm.
Currently the sin and cos calculations for rotation to pitch and roll are only performed at 10Hz.
If I move them out, and upsample to 100Hz, then do the rotation at 100Hz, we are doing that maths a lot more often.
The filter is not adding 'significant' lag, in the sense that it only tracks the corners of the steps, so lag is always less than 100ms.
The downside of the current situation is that there could be 10Hz ripple on pitch or roll, but only while yawing at the same time as the controller was commanding some meaningful amount of pitch or roll.  Otherwise, if there is little pitch or roll, the yaw input would be very smooth.
Perhaps I should make a note here and leave as is?

Reply:
Do not handle it now, but it should be visited later:

> The filter is not adding 'significant' lag, in the sense that it only tracks the corners of the steps, so lag is always less than 100ms.

100ms delay will translate to 36degree direction error if quad is yawing at 1 rotation per second. Such error could lead to circling around  target location, possibly with increasing speed and amplitude (it will definitely happen at 90 deg error).  It is not a problem now, because yaw input is not allowed. 
But stick-controlled yaw during POSHOLD is quite reasonable - allow pilot to set desired heading.

> If I move them out, and upsample to 100Hz, then do the rotation at 100Hz, we are doing that maths a lot more often.

Math is cheap with FPU. It may be noticeable at 8kHz, but at IMU rate, it should be negligible.
Also you can extract 2d rotation matrix from orientation matrix or quaternion, at cost of vector normalization (untested, concept only)
```
    // (chatgpt)
    float R11 = R[0][0];
    float R21 = R[1][0];

    // Compute the norm for R11 and R21
    float norm = sqrtf(R11 * R11 + R21 * R21);

    // Normalize components for the Z-axis rotation matrix
    float cos_theta = R11 / norm;
    float sin_theta = R21 / norm;